Added .odoignore functionality and glob matching support for watch.

mik-dass opened the pull request:
fixes #1036 

Added the support for ** and other patterns for glob matching.

Adds support for reading .odoignore file for getting the ignore rules

A function 'GetIgnoreRulesFromDirectory' is introduced to get the ignore rules from the source directory. If .odoignore is not found, .gitignore will be used. If both files don't exist, empty array would be returned.

NOTE : Currently it assumes the .odoignore or .gitignore to be in the root directory as stated in the issue

mik-dass commented:
@girishramnani Yes I will put that in a separate PR

syamgk commented on pkg/component/watch_test.go:
```diff
@@ -16,41 +16,50 @@ import (
+		strToMatch    string
+		regExps       []string
+		want          bool
+		wantErr       bool
```
unit test is not covering the `return false, err` in https://github.com/redhat-developer/odo/pull/1158/files#diff-08eb53b6a5d51aab5d95b32c1ebcead5R50

mik-dass replied:
According to my observation, `glob.Compile()` doesn't throw error when a wrong expression is passed. So this might be tricky to reproduce in test

kadel replied:
if that is true, than it might be a bug in the library. Why would it return an error?

kadel replied:
You can make it  fail with `[`

mik-dass replied:
Good, thanks :+1:

kadel commented on pkg/component/watch.go:
```diff
@@ -58,7 +62,7 @@ func isRegExpMatch(strToMatch string, regExps []string) (bool, error) {
 // Files matching glob pattern defined in ignores will be ignored.
 // Taken from https://github.com/openshift/origin/blob/85eb37b34f0657631592356d020cef5a58470f8e/pkg/util/fsnotification/fsnotification.go
-func addRecursiveWatch(watcher *fsnotify.Watcher, path string, ignores []string) error {
+func addRecursiveWatch(watcher *fsnotify.Watcher, directory, path string, ignores []string) error {
```
how are  `directory` and `path` different?

kadel replied:
can we use a little bit more descriptive name for `directory` and `path` to make it a little bit more apparent what is what?

kadel commented on pkg/component/watch.go:
```diff
@@ -255,7 +259,7 @@ func WatchAndPush(client *occlient.Client, out io.Writer, parameters WatchParame
 		}
 	}()
-	err = addRecursiveWatch(watcher, parameters.Path, parameters.FileIgnores)
+	err = addRecursiveWatch(watcher, parameters.Path, parameters.Path, parameters.FileIgnores)
```
why is there `parameters.Path` twice?

kadel commented on pkg/component/watch_test.go:
```diff
@@ -16,41 +16,50 @@ import (
+		testName      string
+		directoryName string
+		strToMatch    string
+		regExps       []string
```
this is not regexp anymore is it?

kadel commented on pkg/util/util_test.go:
```diff
@@ -560,3 +561,85 @@ func TestCheckPathExists(t *testing.T) {
+		wantRules     []string
+		wantErr       bool
+	}{
+		{
```
all the tests here are testing either `odoignore` or `gitignore`, but there is no case where both files are present.

kadel replied:
we also need a test case that negates the pattern using `!`

kadel commented:
lines in odoignore and git ignore that are starting with `#` should be ignored
see https://git-scm.com/docs/gitignore#_pattern_format

```
I0103 13:41:27.185069   34910 watch.go:161] starting WatchAndPush, path: /Users/tkral/tmp/odo-examples/spring-boot-http-booster, component: spring-boot-redhat-openjdk18-openshift-kvwh, ignores [.git # IntelliJ .idea *.iml target # Eclipse .metadata .settings .classpath .project bin tmp]
```

kadel commented:
I have the following .gitignore
```

# IntelliJ
.idea
*.iml
target

# Eclipse
.metadata
.settings
.classpath
.project
bin
tmp
```

Watch is not working at all :-( Regardles of what file I chage, nothing happens :-(
Maybe it is ignoring everything?

It would be good to add debug message that informs when files is being ignored becoase of ignore pattern

```
▶ odo watch -v 4
I0103 13:46:36.314797   35796 occlient.go:395] Trying to connect to server 192.168.64.11:8443
I0103 13:46:36.315763   35796 occlient.go:403] Server https://192.168.64.11:8443 is up
I0103 13:46:36.581560   35796 occlient.go:324] isLoggedIn err:  <nil>
 output: "developer"
I0103 13:46:36.582015   35796 context.go:203] Ignoring error as it usually means flag wasn't set: flag accessed but not defined: component
I0103 13:46:36.582341   35796 watch.go:48] No component name passed, assuming current component
I0103 13:46:36.583040   35796 occlient.go:2484] Getting DeploymentConfig: spring-boot-redhat-openjdk18-openshift-kvwh-myapp
I0103 13:46:36.724353   35796 component.go:540] Source for component spring-boot-redhat-openjdk18-openshift-kvwh is file:///Users/tkral/tmp/odo-examples/spring-boot-http-booster (local)
I0103 13:46:36.724841   35796 watch.go:161] starting WatchAndPush, path: /Users/tkral/tmp/odo-examples/spring-boot-http-booster, component: spring-boot-redhat-openjdk18-openshift-kvwh, ignores [.git # IntelliJ .idea *.iml target # Eclipse .metadata .settings .classpath .project bin tmp]
I0103 13:46:36.725542   35796 watch.go:109] ignoring watch on path /Users/tkral/tmp/odo-examples/spring-boot-http-booster/.git
I0103 13:46:36.726505   35796 watch.go:109] ignoring watch on path /Users/tkral/tmp/odo-examples/spring-boot-http-booster/.settings
I0103 13:46:36.731693   35796 watch.go:109] ignoring watch on path /Users/tkral/tmp/odo-examples/spring-boot-http-booster/target
I0103 13:46:36.733756   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster
I0103 13:46:36.735146   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/.circleci
I0103 13:46:36.736671   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/.mvn
I0103 13:46:36.737235   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/.mvn/wrapper
I0103 13:46:36.737448   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/.openshiftio
I0103 13:46:36.737782   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/.vscode
I0103 13:46:36.738165   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src
I0103 13:46:36.738510   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/main
I0103 13:46:36.738873   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/main/java
I0103 13:46:36.739224   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/main/java/io
I0103 13:46:36.739597   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/main/java/io/openshift
I0103 13:46:36.739984   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/main/java/io/openshift/booster
I0103 13:46:36.740357   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/main/java/io/openshift/booster/service
I0103 13:46:36.740721   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/main/resources
I0103 13:46:36.741249   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/main/resources/static
I0103 13:46:36.742106   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/test
I0103 13:46:36.742554   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/test/java
I0103 13:46:36.743028   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/test/java/io
I0103 13:46:36.743373   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/test/java/io/openshift
I0103 13:46:36.744837   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/test/java/io/openshift/booster
I0103 13:46:36.745285   35796 watch.go:126] ignoring watch for /Users/tkral/tmp/odo-examples/spring-boot-http-booster/src/test/resources
Waiting for something to change in /Users/tkral/tmp/odo-examples/spring-boot-http-booster

```

mik-dass commented:
> ines in odoignore and git ignore that are starting with # should be ignored

It seems the library does that automatically. I will just add a test for that

> Watch is not working at all :-( Regardless of what file I change, nothing happens :-(
Maybe it is ignoring everything?

I will have a look and also will add debug messages. Y

mik-dass commented:
@kadel Watch is working fine for me for the repo https://github.com/snowdrop/spring-boot-http-booster

```
[mrinaldas@localhost odo]$ odo watch -v 4
I0104 15:05:18.199213     327 occlient.go:395] Trying to connect to server 127.0.0.1:8443
I0104 15:05:18.199375     327 occlient.go:403] Server https://127.0.0.1:8443 is up
I0104 15:05:18.207410     327 occlient.go:324] isLoggedIn err:  <nil> 
 output: "developer"
I0104 15:05:18.207613     327 context.go:203] Ignoring error as it usually means flag wasn't set: flag accessed but not defined: component
I0104 15:05:18.207739     327 watch.go:48] No component name passed, assuming current component
I0104 15:05:18.207864     327 occlient.go:2484] Getting DeploymentConfig: spring-boot-openjdk18-hptf-odo-hvxr
I0104 15:05:18.214683     327 component.go:540] Source for component spring-boot-openjdk18-hptf is file:///home/mrinaldas/spring-boot-http-booster (local)
I0104 15:05:18.214732     327 watch.go:167] starting WatchAndPush, path: /home/mrinaldas/spring-boot-http-booster, component: spring-boot-openjdk18-hptf, ignores [.git # IntelliJ .idea *.iml target # Eclipse .metadata .settings .classpath .project bin tmp]
I0104 15:05:18.214931     327 watch.go:57] ignoring path /home/mrinaldas/spring-boot-http-booster/.git because of glob rule <text:`/home/mrinaldas/spring-boot-http-booster/.git`>
I0104 15:05:18.214943     327 watch.go:115] ignoring watch on path /home/mrinaldas/spring-boot-http-booster/.git
I0104 15:05:18.215912     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.215926     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster
I0104 15:05:18.216021     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.216026     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/.circleci
I0104 15:05:18.216111     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.216116     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/.mvn
I0104 15:05:18.216195     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.216199     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/.mvn/wrapper
I0104 15:05:18.216275     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.216280     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/.openshiftio
I0104 15:05:18.216356     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.216361     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src
I0104 15:05:18.216441     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.216445     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/main
I0104 15:05:18.216524     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.216529     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/main/java
I0104 15:05:18.216886     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.216910     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/main/java/io
I0104 15:05:18.217169     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.217185     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/main/java/io/openshift
I0104 15:05:18.217394     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.217410     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/main/java/io/openshift/booster
I0104 15:05:18.217617     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.217633     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/main/java/io/openshift/booster/service
I0104 15:05:18.218048     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.218062     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/main/resources
I0104 15:05:18.218374     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.218379     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/main/resources/static
I0104 15:05:18.218480     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.218485     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/test
I0104 15:05:18.218583     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.218588     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/test/java
I0104 15:05:18.218725     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.218729     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/test/java/io
I0104 15:05:18.218873     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.218878     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/test/java/io/openshift
I0104 15:05:18.219049     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.219056     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/test/java/io/openshift/booster
I0104 15:05:18.219204     327 util.go:346] path /home/mrinaldas/spring-boot-http-booster doesn't exist, skipping it
I0104 15:05:18.219208     327 watch.go:141] adding watch on path /home/mrinaldas/spring-boot-http-booster/src/test/resources
Waiting for something to change in /home/mrinaldas/spring-boot-http-booster
```

kadel commented:
@syamgk @girishramnani @surajnarwade have you faced the same issue when you tested this?

kadel commented on pkg/component/watch.go:
```diff
@@ -39,14 +41,18 @@ type WatchParameters struct {
+// Parameters: the component directory (for converting relative path expressions to absolute ones), a string strToMatch and a list of regexp patterns to match strToMatch with
 // Returns: true if there is any match else false
-func isRegExpMatch(strToMatch string, regExps []string) (bool, error) {
+func isRegExpMatch(directory, strToMatch string, regExps []string) (bool, error) {
```
This should have a different name as it is no longer a RegExp

kadel replied:
the same for function comment

kadel commented:
you are still using regexp on line 125 in `pkg/component/watch.go`, shouldn't it be `isRegExpMatch`?

https://github.com/redhat-developer/odo/pull/1158/files#diff-08eb53b6a5d51aab5d95b32c1ebcead5R125

kadel commented:
This issue also proves, that we need e2e tests for this.

mik-dass commented:
> This issue also proves, that we need e2e tests for this.

Sure will have a look into the watch e2e for getting started with it

kadel commented on pkg/component/watch.go:
```diff
@@ -58,7 +64,10 @@ func isRegExpMatch(strToMatch string, regExps []string) (bool, error) {
 // Taken from https://github.com/openshift/origin/blob/85eb37b34f0657631592356d020cef5a58470f8e/pkg/util/fsnotification/fsnotification.go
-func addRecursiveWatch(watcher *fsnotify.Watcher, path string, ignores []string) error {
+// directory is the name of the component source directory
+// path is the path of the file
```
`path` can be also directory!

kadel commented on pkg/component/watch.go:
```diff
@@ -110,16 +119,15 @@ func addRecursiveWatch(watcher *fsnotify.Watcher, path string, ignores []string)
-			}
+
+		if matched, _ := isGlobExpMatch(directory, folder, ignores); matched {
+			ignore = true
```
we don't need `ignore` variable anymore.
It should be enough to move `glog` and `continue` from below `if` statement here.
I don't see `ignore` used anywhere else

kadel commented on pkg/odo/cli/component/watch.go:
```diff
@@ -74,6 +74,14 @@ var watchCmd = &cobra.Command{
+			odoutil.CheckError(err, "")
+		}
+		if len(rules) > 0 {
+			ignores = append(ignores, rules...)
```
Isn't this going to concatenate rules provided as flag and rules from files?
If `--ignore` flag is provided then `odoignore` and `gitignore` should be ignored

mik-dass commented:
@kadel Fixed. Please have a look again :)

anmolbabu commented on pkg/component/watch.go:
```diff
@@ -38,16 +39,21 @@ type WatchParameters struct {
-// Parameters: a string strToMatch and a list of regexp patterns to match strToMatch with
+// isGlobExpMatch compiles strToMatch against each of the passed regExps
+// Parameters: the component directory (for converting relative path expressions to absolute ones), a string strToMatch and a list of regexp patterns to match strToMatch with
 // Returns: true if there is any match else false
```
and error also?

mik-dass replied:
Fixed

anmolbabu commented on pkg/component/watch.go:
```diff
@@ -38,16 +39,21 @@ type WatchParameters struct {
 		}
+		matched := pattern.Match(strToMatch)
 		if matched {
+			glog.V(4).Infof("ignoring path %s because of glob rule %s", strToMatch, globExp)
```
Can this be done in the place where we invoke this function?
Also, this function could be generic right why treat it as a function for only path matching?
Can we pass the path as a parameter in entirety to this function instead of doing the concatenation in this function?
I know at the moment this is like exclusive for push and watch so I'll leave it to you

mik-dass commented on pkg/component/watch.go:
```diff
@@ -38,16 +39,21 @@ type WatchParameters struct {
-// isRegExpMatch compiles strToMatch against each of the passed regExps
-// Parameters: a string strToMatch and a list of regexp patterns to match strToMatch with
+// isGlobExpMatch compiles strToMatch against each of the passed regExps
+// Parameters: the component directory (for converting relative path expressions to absolute ones), a string strToMatch and a list of regexp patterns to match strToMatch with
```
I just modified and followed how it was written earlier

mik-dass replied:
fixed

kadel commented:
This looks good.
But as I was testing this I found one think:
If  .gitignore is used, than we don't ignore `.git`. 

Would it make sense to always ignore `.git`?

amitkrout commented:
@mik-dass Just want to bring some of your attention on the decreasing coverage of file ```pkg/component/watch.go``` https://github.com/redhat-developer/odo/pull/1158#issuecomment-449391710 though we have overall coverage increase. You can give a try to improve it when you have time.

mik-dass commented:
> Would it make sense to always ignore .git?

It seems that .git is automatically ignored so I guess we should too. Will modify the PR accordingly

mik-dass commented:
> @mik-dass Just want to bring some of your attention on the decreasing coverage of file pkg/component/watch.go #1158 (comment) though we have overall coverage increase. You can give a try to improve it when you have time.

@amitkrout I didn't write most of the functions in the file. I just modified some parts according to the PR and also added tests for those parts. I mostly have no idea what rest of the code does :)

amitkrout commented:
> I didn't write most of the functions in the file. I just modified some parts according to the PR and also added tests for those parts. I have no idea how to mock and write tests for the other functions :D

Yup, i understand. That is why i mentioned ```please have a look when you have time``` ;) and more over this is not blocking the pr to merge as part of UT is concerned (The overall coverage is positive).

mik-dass commented:
> Yup, i understand. That is why i mentioned please have a look when you have time ;) and more over this is not blocking the pr to merge as part of UT is concerned (The overall coverage is positive).

@amitkrout Ok cool :+1:  I will have a look into the tests soon. Please help me regarding the same :D

amitkrout commented:
> Please help me regarding the same :D

@mik-dass Ack

mik-dass commented:
@kadel Fixed. Now .git is ignored even if .gitignore or .odoignore is present in the directory.

mik-dass commented:
> @mik-dass Can you please add a user facing scenario in the pr description area (Just a reference for those who want to test this pr)

Done

kadel commented:
one last thing `odo watch -h` still says "regular expressions"
```
      --ignore stringSlice   Files or folders to be ignored via regular expressions.
```

Otherwise LGTM

mik-dass commented:
> one last thing: odo watch -h still says "regular expressions"

Done
